README
What's this ?
-------------
This package uses modular arithmetic to compute exact sign of determinants, and
more general polynomial expressions.  Currently, the entries must have integer
values <2^54, stored into "double"s.  It can be ported to any kind of integer
type, and aligned floating point type.  The main implementation choice is the
one of the moduli, which are chosen prime, down from 2^27, so that the basic
operation "a*d-b*c" fits exactly in a double.

Author:
-------
Sylvain Pion <Sylvain.Pion@sophia.inria.fr> [Nov 96 - Dec 97].
Theoretical work has been done with Herve Bronnimann, Ioannis Emiris and
Victor Pan.  A paper at ACM CG'97 describes that part, and also a RR.
See http://www.inria.fr/prisme/personnel/pion/progs/modular/modular_eng.html
    (or near the URL http://www.inria.fr/prisme).
This work is supported by the CGAL Project.

Copyright:
----------
This package is copyrighted by Sylvain Pion 1997-1998.
This code is placed under the LGPL, see file COPYING.LIB.

Supported platforms:
--------------------
The modular package relies on IEEE 754 FP specifications, and especially some
rounding modes, to compute exact upper bounds of expressions, and setting the
precision to "double" exactly.  So the only file that needs to be ported is
_FPU.h.  Currently, the supported architectures are:
- Sparc/SunOS   (supported with gcc only)
- Sparc/Solaris
- Intel x86/Linux
- Mips/Irix
- Sparc/Linux
- Alpha/Linux   (not yet fully supported)
- Alpha/DEC-OSF (not yet fully supported)
- Please contact me for other ports, and see the file _FPU.h.


How to compile and install:
---------------------------
- Check the compile options in the Makefile.
- Eventually change the sign reconstruction method for determinants, in
  MOD_det.h.
- Type make, you'll get libMOD.a and libMOD.so.
- You can just keep the *.h files and the libraries.


How to use:
-----------
- In your source files, add the line: #include "MOD.h" (and "MOD_det.h" for
  determinants stuff).
- Use the functions:
    int MOD_sign_det    (double **mat, const int dim);
    int MOD_det_is_null (double **mat, const int dim);
  (mat is a pointer -> array of pointers -> arrays of doubles)
- Modular arithmetic requires some precomputations.  So you need to call, at
  the beginning of your program MOD_init(nb_mod) and MOD_det_init(dim).  If
  these numbers are too low, the functions MOD_det_needs_mods() and
  MOD_fp_needs_mods() will take care of that dynamicaly.

  [ The parameters of MOD_det_init() and MOD_init() are the maximum dimension
  you plan to use, and the maximum number of modulis you plan to use. You can
  estimate this number, by dividing the logarithm (base 2) of the maximum
  value expected for your determinant, by 27. ]

- Link with libMOD.a (or libMOD.so) and -lm.
- If you want to use the package at a lower level (to compute different
  expressions than signs of determinants), you can look at MOD_det.c,
  and see how it works.  Mainly, you can use the functions in MOD_base_ops.h,
  after having set MOD_current and MOD_current_inv, which are global.
- Thread support is still experimental.  See the Makefile for enabling that.


Organization:
-------------
- The CHANGELOG file contains all that has been modified since the beginning,
  and the number of the current version.
- The TODO file contains things that aren't implemented yet, but may be in
  next versions.
- MOD.h is the general include file
- MOD_det.h is the same for determinants.
- ...


CHANGELOG
Version 2.4 -- Wed Feb 4 1998
- Fixed bug in MOD_det_init() that supposed enough moduli were precomputed.
  (thanks Fabrice for the report)
- MOD_reconstruct.h split into .c and .h.

Version 2.3 -- Tue Dec 16 1997

- Functions dealing with several residues from MOD_base_ops.h put in
  MOD_reconstruct.h.
- Thread support added (parallel computation on multi-processor machines).
  [ This has implied that a preprocessing is done when MOD_THREAD is not set. ]
- MOD_{sign_det|det_is_null}() take now a "double **mat", not "MOD_type **mat".
- All the determinant stuff is moved into MOD_det.[ch].
- MOD_is_prime.[ch] cleanup.
- MOD_pre_primes[] is now an array of int => space saved.
- MOD_int_TYPE, MOD_type_name, MOD_TYPE_name: removed.
- time_sign_rec program added, to test and bench the different sign
  reconstruction methods.
- MOD_det_clear() and MOD_det_init(dim) must be called separately if you use
  determinants.
- Changed the copyright notice to LGPL.
- MOD_to_double() function added, that gives a good double approximation,
  based on Newton's representation.
- MOD_newton_i[] changed to MOD_newton_coef[].

Version 2.2 -- Sat Nov 15 1997

- MOD_det5x5() removed: useless and took place.
- More dynamic: more moduli are "precomputed" on demand by MOD_needed_mods()
  and MOD_fp_needed_mods().
- Various fixes in MOD_init().
- Added some #ifndef MOD_???_H to protect the include files from re-inclusions.
- MOD_needed_mods()    renamed MOD_det_needs_mods().
- MOD_fp_needed_mods() renamed MOD_fp_needs_mods().
- MOD_modulis[]        renamed MOD_residues[].

Version 2.1 -- Wed Nov 12 1997

- Changed MOD_ABS() from "<?-:" to fabs().
- MOD_det5x5() now done separetely (it's not Gauss anymore).
  It's disabled, because non tested (cf MOD_det.c).
- MOD_perm2(a,b,c,d) = a*d+b*c introduced.
- MOD_rounding.h replaced by _FPU.h (common with the DIS package).
- "long double"s are not used => away from the source, it's cleaner.
- No longer system dependant file, except _FPU.h.


Version 2.0 -- Wed Apr 16 1997

- Big reorganisation... It should be a cleaner package now.
- changed back __inline__ to inline (it was only usefull for ANSI check...)
- main_det.[ch] + gauss.[ch] -> MOD_det.[ch]
- all symbol and file names begin now with MOD_, to avoid name clashes.
- Shared library libMOD.so available.

Version 1.9 -- Wed Mar 12 1997

- syst.h becomes MOD_syst.h, and contains compile-time configurable options.
- Possibility to *not* use the "long double" type under Linux/*86.
  And I made it the default, since it's quicker and takes less place.
- Added null determinant matrices to generate.c.
- MOD_det_is_null() high level function added, that returns (det == 0).
- Reduce memory space needed by some precomputations.
- Merged a few ridiculous files, into main ones.
- libmod.a becomes libMOD.a, and a few other name changes.
- Dynamic test for good primes for long double should be ok now.

Version 1.8 -- Mon Feb 24 1997

- deduction(i) function implemented. You can deduce dets_i[i] from the
  newton_i[0 -> i-1] (which can be computed from dets_i[0 -> i-1]).
  [That's true if det < product (MOD_primes[0 -> i-1]), of course !]
- Fix in the Makefile, so that "make -j" works (compile on multi-processors).
- Use "CFLAGS += -DNEWTON=??" in the Makefile, to use the Newton method.
  You can set the probability in the file main_det.c.
- Sparc/Linux port... in progress.
- Various fixes, so that it compiles with gcc, g++, cc, CC (solaris at least).
  The code is now "ANSI C/C++", with optionnal use of inline functions.

Version 1.7 -- Sun Feb 2 1997

- Newton method, and probabilistic variant.

Version 1.6 -- Wed Jan 29 1997

- No more systematicaly print information in MOD_init().
- Alpha/DEC-OSF port.
- Alpha/Linux port.

Version 1.5 -- Tue Jan 14 1997

- No more limit on the dimension.
- Bug fix in epsilon_lagrange[].
- Pre-computation of epsilon_orig[].
- primes.c, primes.h and primes are back. Prime numbers are computed at compile
  time. If MOD_init() is called with more modulis than those precomputed,
  the one needed are computed at run time. This is a good compromise.

Version 1.4 -- Fri Jan 10 1997

- Exact implementation of Lagrange's method.
- Exact implementation of the "last" method (relax_mul).
- bug fix about the definition of MANTISSA_ERROR.
- Switch to specialized functions in Gauss algorithm, as soon as dim<5.
- MOD_init(dim,#mod) and MOD_clear() functions.
- primes.c, compute_bounds.c are removed (and the files they generated:
  modulis.h, det_bounds.h, primes.o, primes, compute_bounds.o, compute_bounds).
  The precomputations are now done at run time, in MOD_init().
- general.h file, to clean the package.
- inv_MOD[] moved to MOD_primes_inv_maj[].
- Precompute the MOD_primes_inv[] = 1/MOD_primes[], to replace the
  divisions "/ MOD_current", by a multiplication "* MOD_current_inv".
  That speeds up things in MOD_det2, MOD_mul, which is great.
  I'm not sure we have the right to do that.
- Compile time option "FIXED_DIM" removed, as it does not speed up anything.

Version 1.3 -- Fri Dec 20 1996

- Improved Bezout: it's not neccesary to use MOD_reduce at each step,
  only a MOD_soft_reduce at the end is necessary.
- det_bounds.h is now generated by compute_bounds.c, so you can use
  dimensions greater than 14. It should be ok for all dimensions <= 24.
- Little change in Gauss function.
- Support for a lot of modulis (no more bounded by 37 (31 on Intel).
  You can set this in the Makefile (default is 100).

Version 1.2 -- Fri Dec 6 1996

- Bezout instead of exponentiation (Fermat) to compute the inverse.
- modulis.h excluded from libmod.a
- inline functions declared in .h files, so they are now inlined :-)
- time_mod program, using generate.c, that computes a matrix with
  a "fixed" determinant, and entrees on b bits.

Version 1.1 -- Wed Dec 4 1996

- Improvements, less useless computations (in Gauss, numer and denom).
- Irix port
- x86/Linux port, with support of 64 bits mantissas (=> 20% gain)

MOD_det.c
/* This file contains 2 main functions MOD_sign_det() and MOD_det_is_null().
 * They return the sign/nullity of the determinant of the matrix of doubles
 * (containing integer coefficients <2^54).
 */

#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#ifdef MOD_THREADS
#include <pthread.h>
pthread_t MOD_threads[MOD_THREADS];
#endif

#include "_FPU.h"
#include "MOD.h"
#include "MOD_det.h"

#ifndef MOD_THREADS
#define MOD_det2x2(a,b,c)	MOD_det2x2(a)
#define MOD_det3x3(a,b,c)	MOD_det3x3(a)
#define MOD_det4x4(a,b,c)	MOD_det4x4(a)
#define MOD_gauss(a,b,c,d)	MOD_gauss(a,b)
#endif

inline MOD_type MOD_det2x2(MOD_type **mat, MOD_type, MOD_type);
MOD_type MOD_det3x3(MOD_type **mat, MOD_type, MOD_type);
MOD_type MOD_det4x4(MOD_type **mat, MOD_type, MOD_type);
MOD_type MOD_gauss (MOD_type **mat, int dim, MOD_type, MOD_type);
int      MOD_det_needs_mods (double **mat, const int dim);

static MOD_type **MOD_mat_tmp, *MOD_mat_work_space;
static MOD_type *MOD_max_li, *MOD_max_co;
static double *MOD_det_bounds;
static int *MOD_nb_dim_mod, MOD_dim_max;
static int MOD_best_det_bounds[]={1,1,2,4,16,48,160,576,4096,14336,73728,327680,2985984,14929920,77635584};


/* One way to compute dynamicaly the number of modulis needed: Hadamard */
/* NB: problem: to compute the number of moduli needed, we need some
       precomputations to be done... so it's incremental (not optimal) */

int MOD_det_needs_mods (double **mat, const int dim)
{
    double tmp, co=MOD_det_bounds[dim], li=co;
    int i, j;
    int nb_col=MOD_nb_dim_mod[dim];
    int nb_lig=MOD_nb_dim_mod[dim];

    if (dim>MOD_dim_max)
    {
    	MOD_dim_max = dim;
    	MOD_det_clear();
	MOD_det_init(MOD_dim_max);
    };

    for (i=0; i<dim; i++)
        MOD_max_li[i] = MOD_max_co[i] = 0;

    for (i=0; i<dim; i++)
        for (j=0; j<dim; j++)
        {
            tmp = MOD_ABS(mat[i][j]);
            MOD_max_li[i] = MOD_MAX(MOD_max_li[i], tmp);
            MOD_max_co[j] = MOD_MAX(MOD_max_co[j], tmp);
        };

    _FPU_set_rounding_to_infinity();

    for (i=0; i<dim; i++)
    {
        li *= MOD_max_li[i];
	while (li*MOD_epsilon_orig[nb_lig] > .5)
        {
	    li *= MOD_primes_inv_maj[nb_lig++];
            if (nb_lig > MOD_nb_mod_max)
            {
              MOD_clear();
	      _FPU_set_rounding_to_nearest();
              MOD_init(MOD_nb_mod_max+10);
              return MOD_det_needs_mods (mat,dim);
            };
        };
        co *= MOD_max_co[i];
	while (co*MOD_epsilon_orig[nb_col] > .5)
        {
	    co *= MOD_primes_inv_maj[nb_col++];
            if (nb_col > MOD_nb_mod_max)
            {
              MOD_clear();
	      _FPU_set_rounding_to_nearest();
              MOD_init(MOD_nb_mod_max+10);
              return MOD_det_needs_mods (mat,dim);
            };
        };
    };

    _FPU_set_rounding_to_nearest();

    if ((li == 0.0) || (co == 0.0))
        return 0;

    return MOD_MIN(nb_col, nb_lig);
}

inline MOD_type MOD_det2x2(MOD_type **mat, MOD_type MOD_current, MOD_type MOD_current_inv)
{
    return MOD_det2(    mat[0][0], mat[0][1],
			mat[1][0], mat[1][1], MOD_current, MOD_current_inv);
}

MOD_type MOD_det3x3(MOD_type **mat, MOD_type MOD_current, MOD_type MOD_current_inv)
{
    MOD_type a = MOD_det2(  mat[1][1], mat[1][2],
			    mat[2][1], mat[2][2], MOD_current, MOD_current_inv);
    MOD_type b = MOD_det2(  mat[0][1], mat[0][2],
			    mat[2][1], mat[2][2], MOD_current, MOD_current_inv);
    MOD_type c = MOD_det2(  mat[0][1], mat[0][2],
			    mat[1][1], mat[1][2], MOD_current, MOD_current_inv);
    MOD_type d = MOD_det2(  mat[0][0], mat[1][0],
			    b, a, MOD_current, MOD_current_inv);
    a = MOD_mul(mat[2][0], c, MOD_current, MOD_current_inv);
    return MOD_add(a, d, MOD_current, MOD_current_inv);
}

MOD_type MOD_det4x4(MOD_type **mat, MOD_type MOD_current, MOD_type MOD_current_inv)
{
    MOD_type p0 = MOD_det2( mat[0][0], mat[1][0],
			    mat[0][1], mat[1][1], MOD_current, MOD_current_inv);
    MOD_type p1 = MOD_det2( mat[0][0], mat[2][0],
			    mat[0][1], mat[2][1], MOD_current, MOD_current_inv);
    MOD_type p2 = MOD_det2( mat[0][0], mat[3][0],
			    mat[0][1], mat[3][1], MOD_current, MOD_current_inv);
    MOD_type p3 = MOD_det2( mat[1][0], mat[2][0],
			    mat[1][1], mat[2][1], MOD_current, MOD_current_inv);
    MOD_type p4 = MOD_det2( mat[1][0], mat[3][0],
			    mat[1][1], mat[3][1], MOD_current, MOD_current_inv);
    MOD_type p5 = MOD_det2( mat[2][0], mat[3][0],
			    mat[2][1], mat[3][1], MOD_current, MOD_current_inv);
    MOD_type q0 = MOD_det2( mat[0][2], mat[1][2],
			    mat[0][3], mat[1][3], MOD_current, MOD_current_inv);
    MOD_type q1 = MOD_det2( mat[0][2], mat[2][2],
			    mat[0][3], mat[2][3], MOD_current, MOD_current_inv);
    MOD_type q2 = MOD_det2( mat[0][2], mat[3][2],
			    mat[0][3], mat[3][3], MOD_current, MOD_current_inv);
    MOD_type q3 = MOD_det2( mat[1][2], mat[2][2],
			    mat[1][3], mat[2][3], MOD_current, MOD_current_inv);
    MOD_type q4 = MOD_det2( mat[1][2], mat[3][2],
			    mat[1][3], mat[3][3], MOD_current, MOD_current_inv);
    MOD_type q5 = - MOD_det2( mat[2][2], mat[3][2], /* This one is opposed. */
			    mat[2][3], mat[3][3], MOD_current, MOD_current_inv);

    /* Then do: p0.(-q5) - p1.q4 + p2.q3 + p3.q2 - p4.q1 + p5.q0. */

    MOD_type a = MOD_det2(p2,p1,q4,q3, MOD_current, MOD_current_inv);
    MOD_type b = MOD_det2(p3,p0,q5,q2, MOD_current, MOD_current_inv);
    MOD_type c = MOD_det2(p5,p4,q1,q0, MOD_current, MOD_current_inv);

    return MOD_add(MOD_add(a,b, MOD_current, MOD_current_inv),c, MOD_current, MOD_current_inv);
}


/* Implementation of the Gauss algorithm, using RNS.
 *
 * NB1: Does not check if dimension == 1 or 0 (maybe crash :-).
 * NB2: Works only if MOD_current is prime.
 */

MOD_type MOD_gauss (MOD_type **mat, int dim, MOD_type MOD_current, MOD_type MOD_current_inv)
{
    switch (dim)
    {
    	case 2:	return MOD_det2x2(mat, MOD_current, MOD_current_inv);
    	case 3:	return MOD_det3x3(mat, MOD_current, MOD_current_inv);
    	case 4:	return MOD_det4x4(mat, MOD_current, MOD_current_inv);
    	default:
	{
	int i, j, pivot;
	MOD_type numer = 1.0;
	MOD_type denom = 1.0;
	MOD_type denom_partial = 1.0;	/* partial product of the pivots. */
	MOD_type pivot_value;		/* caches the value of the pivot. */

	/* Gauss is "recursif terminal", so I prefer a loop. */

	while (dim != 4)
	{
	    /* looking for the pivot. */

	    pivot = 0;
	    while (( pivot_value = mat[pivot][0] ) == 0)
	    {
		if ((++pivot) == dim) return 0;
		denom = -denom;
	    };

	    denom_partial = MOD_mul(denom_partial, pivot_value, MOD_current, MOD_current_inv);
	    denom = MOD_mul(denom, denom_partial, MOD_current, MOD_current_inv);

	    for (i=0; i<dim; i++)
		if (i != pivot)
		    for (j=1; j<dim; j++)
			mat[i][j] = MOD_det2(   pivot_value, mat[pivot][j],
						mat[i][0]  , mat[i][j], MOD_current, MOD_current_inv);
	    --dim;
	    for (i=0; i<dim; i++)
		mat[i] = 1 + mat[i + (i>=pivot)];
	};

	numer = MOD_det4x4(mat, MOD_current, MOD_current_inv);
	denom_partial = MOD_mul(denom_partial,denom_partial, MOD_current, MOD_current_inv);
	denom = MOD_mul (denom, denom_partial, MOD_current, MOD_current_inv);

	/* Returns the result of the final division. */
	return MOD_div (numer, denom, MOD_current, MOD_current_inv);
	};
    };
}

#ifdef MOD_THREADS
typedef struct {
  int thr, dim;
  MOD_type **mat;
} pass_thr;

void *MOD_gauss_thr (void *pass_arg)
{
  int i,j,k;
  int thr = ((pass_thr *) pass_arg)->thr;
  int dim = ((pass_thr *) pass_arg)->dim;
  MOD_type **mat = ((pass_thr *) pass_arg)->mat;
  MOD_type MOD_current, MOD_current_inv;

  for (i=thr; i<MOD_nb_modulis; i+=MOD_THREADS)
  {
    MOD_mat_tmp[thr*MOD_THREADS] = MOD_mat_work_space + dim*dim*thr;
    for (j=1+thr*MOD_THREADS; j<dim+thr*MOD_THREADS; j++)
      MOD_mat_tmp[j] = MOD_mat_tmp[j-1] + dim;
    MOD_current = MOD_primes[i];
    MOD_current_inv = MOD_primes_inv[i];

    for (j=0; j<dim; j++)
      for (k=0; k<dim; k++)
        MOD_mat_tmp[j+thr*MOD_THREADS][k] = MOD_reduce(mat[j][k], MOD_current, MOD_current_inv);

    MOD_residues[i] = MOD_gauss (MOD_mat_tmp + thr*MOD_THREADS, dim, MOD_current, MOD_current_inv);

// Sometimes it works, sometimes it doesn't...  I disbeleive it !!!

  };
  return NULL;
}
#endif

int MOD_sign_det (double **mat, const int dim)
{
#ifndef MOD_THREADS
  int i,j,k;
#else
  int thr;
  pass_thr tab_thr[MOD_THREADS];
#endif
#ifdef MOD_NEWTON
  int newton_sig=0;
  int newton_current_prob = MOD_NEWTON;
#endif

/* Determine a bound on the deterninant => number of modulis needed. */

  MOD_nb_modulis = MOD_det_needs_mods(mat,dim);
  if (MOD_nb_modulis == 0) return 0;

/* Computes all (det % m_i). */

#ifndef MOD_THREADS
  for (i=0; i<MOD_nb_modulis; i++)
  {
    MOD_mat_tmp[0] = MOD_mat_work_space;
    for (j=1; j<dim; j++)
      MOD_mat_tmp[j] = MOD_mat_tmp[j-1] + dim;
    MOD_current = MOD_primes[i];
    MOD_current_inv = MOD_primes_inv[i];

    /* Compute the matrix of coefs % MOD_current */
    for (j=0; j<dim; j++)
      for (k=0; k<dim; k++)
	MOD_mat_tmp[j][k] = MOD_reduce(mat[j][k], MOD_current, MOD_current_inv); 
    /* Use Gauss to compute the determinant modulo MOD_current. */
    MOD_residues[i] = MOD_gauss (MOD_mat_tmp,dim, MOD_current, MOD_current_inv);

#ifdef MOD_NEWTON
    /* Compute Newton's coefficient (only useful for Newton's method). */
    MOD_newton(i);
    if (MOD_newton_coef[i] != 0)
    {
      newton_current_prob = MOD_NEWTON;
      newton_sig = (MOD_newton_coef[i] > 0) ? 1 : -1;
    }
    else	/* Probabilistic variant. */
    if ((--newton_current_prob) == 0)
      return newton_sig;
#endif /* MOD_NEWTON */
  };
#else /* MOD_THREADS */
#ifdef MOD_NEWTON
#  error "Newton is not compatible with Threads. Sorry !"
#endif

	/* Let the threads go ! */
  for (thr=0; thr<MOD_THREADS; thr++)
  {
    tab_thr[thr].thr = thr;
    tab_thr[thr].dim = dim;
    tab_thr[thr].mat = mat;
    pthread_create(&MOD_threads[thr], NULL, *MOD_gauss_thr, (void *) (tab_thr + thr));
  };

        /* Wait for the threads to finish. */
  for (thr=0; thr<MOD_THREADS; thr++)
    pthread_join(MOD_threads[thr],NULL);

#endif /* MOD_THREADS */

    /* Computes the sign. */
#ifdef MOD_NEWTON
  return newton_sig;
#endif
#ifdef MOD_LAGRANGE
  return MOD_lagrange();
#endif
#ifdef MOD_RELAX_MUL
  return MOD_relax_mul();
#endif
}


/* Same, but check only if ==0 or not.
 * It's far quicker in average, if det !=0.
 * Be carefull, it returns MOD_TRUE or MOD_FALSE, not 0 or 1,
 * whether det==0 or not.
 */

int MOD_det_is_null (double **mat, const int dim)
{
    int i,j,k;
#ifdef MOD_THREADS
    MOD_type MOD_current, MOD_current_inv;
#endif

/* Determine a bound on the deterninant => number of modulis needed. */

    MOD_nb_modulis = MOD_det_needs_mods(mat,dim);
    if (MOD_nb_modulis == 0) return 0;

/* Computes all (det % m_i). */

    for (i=0; i<MOD_nb_modulis; i++)
    {
        MOD_mat_tmp[0] = MOD_mat_work_space;
	for (j=1; j<dim; j++)
	    MOD_mat_tmp[j] = MOD_mat_tmp[j-1] + dim;
	MOD_current = MOD_primes[i];
	MOD_current_inv = MOD_primes_inv[i];
    /* Compute the matrix of coefs % MOD_current */
	for (j=0; j<dim; j++)
	    for (k=0; k<dim; k++)
		MOD_mat_tmp[j][k] = MOD_reduce(mat[j][k], MOD_current, MOD_current_inv);
    /* Use Gauss to compute the determinant modulo MOD_current.
     * And end if possible.
     */
	if (MOD_gauss (MOD_mat_tmp,dim, MOD_current, MOD_current_inv) != 0) return MOD_FALSE;
    };
    return MOD_TRUE;	/* All det \mod m_i are null. */
}


void MOD_det_init(int dim)
{
    int i, j;

    MOD_dim_max = dim;

    /* Reserve some memory to compute the determinants modulo m_i.
     * MOD_mat_work_space[] and MOD_mat_tmp[].
     */

#ifdef MOD_THREADS
    MOD_mat_work_space = (MOD_type *) malloc(sizeof(MOD_type) *MOD_dim_max*MOD_dim_max*MOD_THREADS);
    MOD_mat_tmp = (MOD_type **) malloc(sizeof(MOD_type *) * MOD_dim_max*MOD_THREADS);
#else
    MOD_mat_work_space = (MOD_type *) malloc(sizeof(MOD_type) *MOD_dim_max*MOD_dim_max);
    MOD_mat_tmp = (MOD_type **) malloc(sizeof(MOD_type *) * MOD_dim_max);
#endif
    if ( (MOD_mat_work_space == NULL) || (MOD_mat_tmp == NULL) ) exit(-1);

    /* Needed by function MOD_det_needs_mods(). */

    MOD_max_li = (MOD_type *) malloc (sizeof(MOD_type) * MOD_dim_max);
    MOD_max_co = (MOD_type *) malloc (sizeof(MOD_type) * MOD_dim_max);
    if ( (MOD_max_li == NULL) || (MOD_max_co == NULL) ) exit(-1);

    _FPU_set_rounding_to_nearest();

    /* Computes the best bounds known for determinants.
     * It's predefined til dimension 14, then Hadamard's bound is used.
     * NB: MOD_nb_dim_mod contains the number of moduli needed,
     * and MOD_det_bounds[] contains the rest quantity.
     */

    MOD_det_bounds = (double *) malloc(sizeof(double) * (MOD_dim_max+1));
    MOD_nb_dim_mod = (int *) malloc(sizeof(int) * (MOD_dim_max+1));
    if ( (MOD_det_bounds == NULL) || (MOD_nb_dim_mod == NULL) ) exit(-1);
    _FPU_set_rounding_to_infinity();

    for (i=0; i<MOD_MIN(sizeof(MOD_best_det_bounds)/sizeof(int),MOD_dim_max+1); i++)
    {
        MOD_det_bounds[i] = MOD_best_det_bounds[i];
        MOD_nb_dim_mod[i] = 0;
    };
    while (i<=MOD_dim_max)
    {
    /* Compute Hadamard's bound, ie d^(d/2). */

        MOD_nb_dim_mod[i] = 0;
        MOD_det_bounds[i] = 1.0;
        for(j=1; j<=(i/2);j++)
        {
            MOD_det_bounds[i] *= i;
            if (MOD_det_bounds[i]/MOD_primes[MOD_nb_dim_mod[i]] > 1.0)
            {
                MOD_det_bounds[i] /= MOD_primes[MOD_nb_dim_mod[i]];
                ++MOD_nb_dim_mod[i];
		if (MOD_nb_dim_mod[i] >= MOD_nb_mod_max)
		{
		    MOD_clear();
		    _FPU_set_rounding_to_nearest();
		    MOD_init(MOD_nb_mod_max+10);
		    _FPU_set_rounding_to_infinity();
		};
            };
        };
        if (i%2 != 0)
        {
            MOD_det_bounds[i] *= sqrt(i);
            if (MOD_det_bounds[i]/MOD_primes[MOD_nb_dim_mod[i]] > 1.0)
            {
                MOD_det_bounds[i] /= MOD_primes[MOD_nb_dim_mod[i]];
                ++MOD_nb_dim_mod[i];
                if (MOD_nb_dim_mod[i] >= MOD_nb_mod_max)
                {
                    MOD_clear();
                    _FPU_set_rounding_to_nearest();
                    MOD_init(MOD_nb_mod_max+10);
                    _FPU_set_rounding_to_infinity();
                };
            };
        };
        ++i;
    };

    _FPU_set_rounding_to_nearest();
}

void MOD_det_clear()
{
    free(MOD_mat_work_space);
    free(MOD_mat_tmp);
    free(MOD_max_li);
    free(MOD_max_co);
    free(MOD_det_bounds);
    free(MOD_nb_dim_mod);
}
